Mathlib/CategoryTheory/FiberedCategory/BasedCategory.lean

```diff
@@ -0,0 +1,259 @@
+    BasedNatTrans F H where
+  toNatTrans := CategoryTheory.NatTrans.vcomp α.toNatTrans β.toNatTrans
+  aboveId' := by
+    intro a
+    rw [CategoryTheory.NatTrans.vcomp_app]
+    infer_instance
+
+@[simp]
+lemma comp_toNatTrans {F G H : BasedFunctor 𝒳 𝒴} (α : BasedNatTrans F G) (β : BasedNatTrans G H) :
+    (comp α β).toNatTrans = NatTrans.vcomp α.toNatTrans β.toNatTrans :=
+  rfl
+
+end BasedNatTrans
+
+namespace BasedCategory
+
+open BasedFunctor BasedNatTrans
+
+@[simps!]
+instance homCategory (𝒳 𝒴 : BasedCategory 𝒮) : Category (BasedFunctor 𝒳 𝒴) where
+  Hom := BasedNatTrans
+  id := BasedNatTrans.id
+  comp := BasedNatTrans.comp
+
+section
+
+variable {𝒳 𝒴 : BasedCategory 𝒮}
+
+@[ext]
+lemma homCategory.ext {F G : BasedFunctor 𝒳 𝒴} (α β : F ⟶ G) (h : α.toNatTrans = β.toNatTrans) :
+    α = β :=
+  BasedNatTrans.ext α β h
+
+/-- The inverse of a based natural transformation whose underlying natural tranformation is an
+isomorphism -/
+def BasedNatIso {F G : BasedFunctor 𝒳 𝒴} (α : F.toFunctor ≅ G.toFunctor)
+    (aboveId' : ∀ a : 𝒳.cat, IsHomLift 𝒴.p (𝟙 (𝒳.p.obj a)) (α.hom.app a)) : F ≅ G where
+  hom := { toNatTrans := α.hom }
+  inv := {
+    toNatTrans := α.inv
+    aboveId' := by
+      intro a
+      specialize aboveId' a
+      rw [← NatIso.app_inv]
+      rw [← NatIso.app_hom] at aboveId'
+      apply IsHomLift.lift_id_inv
+  }
+
+/-- The inverse of a based natural transformation whose underlying natural tranformation carries an
+`IsIso` instance. -/
+noncomputable def BasedNatIso_of_isIso {F G : BasedFunctor 𝒳 𝒴} (α : F.toFunctor ⟶ G.toFunctor)
+    [IsIso α] (aboveId' : ∀ a : 𝒳.cat, IsHomLift 𝒴.p (𝟙 (𝒳.p.obj a)) (α.app a)) : F ≅ G where
+  hom := { toNatTrans := α }
+  inv := { toNatTrans := inv α, aboveId' := fun a => by simp [lift_id_inv_isIso] }
+
+/-- The identity natural transformation is a based natural isomorphism -/
+@[simps]
+def BasedNatIso.id (F : BasedFunctor 𝒳 𝒴) : F ≅ F where
+  hom := 𝟙 F
+  inv := 𝟙 F
+
+/-- Left-whiskering in the bicategory `BasedCategory` is given by whiskering the underlying functors
+and natural transformations -/
+@[simps!]
+def whiskerLeft {𝒵 : BasedCategory 𝒮} (F : BasedFunctor 𝒳 𝒴) {G H : BasedFunctor 𝒴 𝒵}
+    (α : G ⟶ H) : BasedFunctor.comp F G ⟶ BasedFunctor.comp F H where
+  toNatTrans := CategoryTheory.whiskerLeft F.toFunctor α.toNatTrans
+  aboveId' := fun a => α.aboveId (F.w_obj a)
+
+/-- Right-whiskering in the bicategory `BasedCategory` is given by whiskering the underlying
+functors and natural transformations -/
+@[simps!]
+def whiskerRight {𝒵 : BasedCategory 𝒮} {F G : BasedFunctor 𝒳 𝒴} (α : F ⟶ G)
+    (H : BasedFunctor 𝒴 𝒵) : BasedFunctor.comp F H ⟶ BasedFunctor.comp G H where
+  toNatTrans := CategoryTheory.whiskerRight α.toNatTrans H.toFunctor
+  aboveId' := fun a => by apply BasedFunctor.pres_IsHomLift
+
+end
+
+/-- `BasedCategory 𝒮` forms a bicategory -/
+instance bicategory : Bicategory (BasedCategory 𝒮) where
+  Hom 𝒳 𝒴 := BasedFunctor 𝒳 𝒴
```

Comment:
I have not worked with bicategories in mathlib, but I think you could define the `CategoryStruct` instance on `BasedCategory 𝒮` first, prove a few lemmas/definitions using this instance (which allow you to use the notations `≫` and `𝟙`), and finally only defining the `Bicategory` instance.